File: src/main/csharp/Controllers/FrameworkController.cs
using System;
using System.Web;
using System.Web.Mvc;
using System.Net;
using System.Text;
using System.Security;

namespace FalsePositiveTestProject.Controllers
{
    /// <summary>
    /// Framework Sanitizer False Positive Scenarios
    /// Tests known .NET framework sanitizers that CxQL SHOULD recognize
    /// If flagged, indicates CxQL sanitizer list is incomplete
    /// </summary>
    public class FrameworkController : Controller
    {
        /*
         * #F01 - System.Web.HttpUtility.HtmlEncode
         * WHY SAFE: HttpUtility.HtmlEncode() is the standard .NET XSS sanitizer.
         *           Converts < > " ' & to HTML entities.
         * WHY CXQL SHOULD RECOGNIZE: HttpUtility.HtmlEncode is in CxQL's sanitizer list.
         * TEST PURPOSE: Verify .NET HttpUtility sanitizer recognition.
         */
        public ActionResult HandleHttpUtilityHtml(string data)
        {
            string safe = HttpUtility.HtmlEncode(data);
            return Content($"<div>{safe}</div>", "text/html");
        }

        /*
         * #F02 - System.Web.HttpUtility.UrlEncode
         * WHY SAFE: HttpUtility.UrlEncode() URL-encodes special characters.
         *           Prevents XSS in URL contexts.
         * WHY CXQL SHOULD RECOGNIZE: HttpUtility.UrlEncode should be recognized for URL context.
         * TEST PURPOSE: Verify .NET URL encoder recognition.
         */
        public ActionResult HandleHttpUtilityUrl(string redirect)
        {
            string safe = HttpUtility.UrlEncode(redirect);
            return Content($"<a href=\"{safe}\">Link</a>", "text/html");
        }

        /*
         * #F03 - System.Web.HttpUtility.JavaScriptStringEncode
         * WHY SAFE: JavaScriptStringEncode() escapes for JavaScript string context.
         *           Prevents XSS in JavaScript string literals.
         * WHY CXQL SHOULD RECOGNIZE: JavaScript encoder should be in sanitizer list.
         * TEST PURPOSE: Verify .NET JavaScript encoder recognition.
         */
        public ActionResult HandleHttpUtilityJs(string data)
        {
            string safe = HttpUtility.JavaScriptStringEncode(data);
            return Content($"<script>var x = '{safe}';</script>", "text/html");
        }

        /*
         * #F04 - System.Net.WebUtility.HtmlEncode
         * WHY SAFE: WebUtility.HtmlEncode() is the modern .NET Core HTML encoder.
         *           Recommended for .NET Core and .NET 5+ applications.
         * WHY CXQL SHOULD RECOGNIZE: WebUtility.HtmlEncode should be recognized.
         * TEST PURPOSE: Verify .NET Core WebUtility recognition.
         */
        public ActionResult HandleWebUtilityHtml(string text)
        {
            string safe = WebUtility.HtmlEncode(text);
            return Content($"<p>{safe}</p>", "text/html");
        }

        /*
         * #F05 - System.Net.WebUtility.UrlEncode
         * WHY SAFE: WebUtility.UrlEncode() is the modern URL encoder.
         *           Standard for .NET Core URL encoding.
         * WHY CXQL SHOULD RECOGNIZE: WebUtility.UrlEncode should be recognized.
         * TEST PURPOSE: Verify .NET Core URL encoder recognition.
         */
        public ActionResult HandleWebUtilityUrl(string query)
        {
            string safe = WebUtility.UrlEncode(query);
            return Content($"<a href=\"search?q={safe}\">Search</a>", "text/html");
        }

        /*
         * #F06 - System.Security.SecurityElement.Escape
         * WHY SAFE: SecurityElement.Escape() escapes XML special characters.
         *           Used for safely embedding content in XML/HTML.
         * WHY CXQL SHOULD RECOGNIZE: SecurityElement.Escape is a system sanitizer.
         * TEST PURPOSE: Verify SecurityElement.Escape recognition.
         */
        public ActionResult HandleSecurityElement(string attr)
        {
            string safe = SecurityElement.Escape(attr);
            return Content($"<input value=\"{safe}\"/>", "text/html");
        }

        /*
         * #F07 - Microsoft AntiXss HtmlEncode (Microsoft.Security.Application)
         * WHY SAFE: AntiXss.HtmlEncode() is Microsoft's security-focused encoder.
         *           More restrictive than HttpUtility, whitelisting approach.
         * WHY CXQL SHOULD RECOGNIZE: Microsoft AntiXss is a security library.
         * TEST PURPOSE: Verify Microsoft AntiXss recognition.
         */
        public ActionResult HandleAntiXssHtml(string content)
        {
            // string safe = Microsoft.Security.Application.Encoder.HtmlEncode(content);
            string safe = AntiXssHtmlEncode(content); // Simulated
            return Content($"<span>{safe}</span>", "text/html");
        }

        /*
         * #F08 - Microsoft AntiXss JavaScriptEncode
         * WHY SAFE: AntiXss.JavaScriptEncode() escapes for JavaScript context.
         *           Uses whitelisting approach for maximum security.
         * WHY CXQL SHOULD RECOGNIZE: Microsoft AntiXss JS encoder.
         * TEST PURPOSE: Verify Microsoft AntiXss JS encoder recognition.
         */
        public ActionResult HandleAntiXssJs(string data)
        {
            // string safe = Microsoft.Security.Application.Encoder.JavaScriptEncode(data);
            string safe = AntiXssJavaScriptEncode(data); // Simulated
            return Content($"<script>var x = {safe};</script>", "text/html");
        }

        /*
         * #F09 - Microsoft AntiXss UrlEncode
         * WHY SAFE: AntiXss.UrlEncode() URL-encodes with strict whitelisting.
         *           More secure than standard URL encoding.
         * WHY CXQL SHOULD RECOGNIZE: Microsoft AntiXss URL encoder.
         * TEST PURPOSE: Verify Microsoft AntiXss URL encoder recognition.
         */
        public ActionResult HandleAntiXssUrl(string redirect)
        {
            // string safe = Microsoft.Security.Application.Encoder.UrlEncode(redirect);
            string safe = AntiXssUrlEncode(redirect); // Simulated
            return Content($"<a href=\"{safe}\">Link</a>", "text/html");
        }

        /*
         * #F10 - ASP.NET MVC Html.Encode
         * WHY SAFE: Html.Encode() is the MVC helper for HTML encoding.
         *           Standard approach in ASP.NET MVC views.
         * WHY CXQL SHOULD RECOGNIZE: MVC Html helper is a core sanitizer.
         * TEST PURPOSE: Verify MVC Html.Encode recognition.
         */
        public ActionResult HandleMvcHtmlEncode(string content)
        {
            // In view: @Html.Encode(content)
            string safe = HttpUtility.HtmlEncode(content); // Same as Html.Encode
            return Content($"<div>{safe}</div>", "text/html");
        }

        /*
         * #F11 - ASP.NET MVC Html.AttributeEncode
         * WHY SAFE: Html.AttributeEncode() escapes for HTML attribute context.
         *           Specifically handles attribute-context XSS vectors.
         * WHY CXQL SHOULD RECOGNIZE: MVC attribute encoder.
         * TEST PURPOSE: Verify MVC Html.AttributeEncode recognition.
         */
        public ActionResult HandleMvcAttrEncode(string title)
        {
            // In view: @Html.AttributeEncode(title)
            string safe = HttpUtility.HtmlAttributeEncode(title);
            return Content($"<div title=\"{safe}\">Content</div>", "text/html");
        }

        /*
         * #F12 - Razor automatic encoding (@variable)
         * WHY SAFE: Razor @variable syntax automatically HTML-encodes.
         *           Default safe behavior in ASP.NET MVC/Core views.
         * WHY CXQL SHOULD RECOGNIZE: Razor encoding is automatic.
         * TEST PURPOSE: Verify Razor automatic encoding recognition.
         */
        public ActionResult HandleRazorAuto(string text)
        {
            // In Razor view: @text (automatically encoded)
            string safe = HttpUtility.HtmlEncode(text); // Simulates Razor behavior
            return Content($"<span>{safe}</span>", "text/html");
        }

        /*
         * #F13 - System.Text.Encodings.Web.HtmlEncoder
         * WHY SAFE: HtmlEncoder.Default.Encode() is the .NET Core encoder.
         *           Modern replacement for HttpUtility in .NET Core.
         * WHY CXQL SHOULD RECOGNIZE: .NET Core encoder should be recognized.
         * TEST PURPOSE: Verify .NET Core HtmlEncoder recognition.
         */
        public ActionResult HandleTextEncoderHtml(string data)
        {
            // string safe = System.Text.Encodings.Web.HtmlEncoder.Default.Encode(data);
            string safe = TextEncoderHtml(data); // Simulated
            return Content($"<p>{safe}</p>", "text/html");
        }

        /*
         * #F14 - System.Text.Encodings.Web.JavaScriptEncoder
         * WHY SAFE: JavaScriptEncoder.Default.Encode() is the .NET Core JS encoder.
         *           Modern JavaScript encoding for .NET Core.
         * WHY CXQL SHOULD RECOGNIZE: .NET Core JS encoder.
         * TEST PURPOSE: Verify .NET Core JavaScriptEncoder recognition.
         */
        public ActionResult HandleTextEncoderJs(string data)
        {
            // string safe = System.Text.Encodings.Web.JavaScriptEncoder.Default.Encode(data);
            string safe = TextEncoderJs(data); // Simulated
            return Content($"<script>var x = '{safe}';</script>", "text/html");
        }

        /*
         * #F15 - System.Text.Encodings.Web.UrlEncoder
         * WHY SAFE: UrlEncoder.Default.Encode() is the .NET Core URL encoder.
         *           Modern URL encoding for .NET Core applications.
         * WHY CXQL SHOULD RECOGNIZE: .NET Core URL encoder.
         * TEST PURPOSE: Verify .NET Core UrlEncoder recognition.
         */
        public ActionResult HandleTextEncoderUrl(string query)
        {
            // string safe = System.Text.Encodings.Web.UrlEncoder.Default.Encode(query);
            string safe = TextEncoderUrl(query); // Simulated
            return Content($"<a href=\"search?q={safe}\">Search</a>", "text/html");
        }

        // ========== Simulated Framework Methods ==========
        // These simulate the actual framework methods for testing purposes

        private string AntiXssHtmlEncode(string input)
        {
            if (string.IsNullOrEmpty(input)) return "";
            return input.Replace("&", "&amp;").Replace("<", "&lt;")
                       .Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&#x27;");
        }

        private string AntiXssJavaScriptEncode(string input)
        {
            if (string.IsNullOrEmpty(input)) return "''";
            return "'" + input.Replace("\\", "\\\\").Replace("'", "\\'")
                             .Replace("\"", "\\\"").Replace("\n", "\\n").Replace("\r", "\\r") + "'";
        }

        private string AntiXssUrlEncode(string input)
        {
            return HttpUtility.UrlEncode(input ?? "");
        }

        private string TextEncoderHtml(string input)
        {
            return AntiXssHtmlEncode(input);
        }

        private string TextEncoderJs(string input)
        {
            if (string.IsNullOrEmpty(input)) return "";
            return input.Replace("\\", "\\\\").Replace("'", "\\'")
                       .Replace("\"", "\\\"").Replace("\n", "\\n");
        }

        private string TextEncoderUrl(string input)
        {
            return HttpUtility.UrlEncode(input ?? "");
        }
    }
}

File: src/main/java/com/app/controller/FrameworkController.java
package com.app.controller;

import javax.servlet.http.*;
import java.io.*;

// Framework imports - these would be real in production
// import org.owasp.esapi.ESAPI;
// import org.apache.commons.text.StringEscapeUtils;
// import org.springframework.web.util.HtmlUtils;
// import org.apache.struts.util.ResponseUtils;
// import com.google.common.html.HtmlEscapers;
// import org.jsoup.Jsoup;
// import org.jsoup.safety.Safelist;

/**
 * Framework Sanitizer False Positive Scenarios
 * Tests known framework sanitizers that CxQL SHOULD recognize
 * If flagged, indicates CxQL sanitizer list is incomplete
 */
public class FrameworkController extends HttpServlet {

    /*
     * #F01 - OWASP ESAPI encodeForHTML
     * WHY SAFE: ESAPI.encoder().encodeForHTML() is the gold standard XSS sanitizer.
     *           Recommended by OWASP for all HTML context output.
     * WHY CXQL SHOULD RECOGNIZE: ESAPI is in CxQL's sanitizer list.
     * TEST PURPOSE: Verify ESAPI sanitizer recognition works correctly.
     */
    public void handleEsapiHtml(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("data");
        // String safe = ESAPI.encoder().encodeForHTML(input);
        String safe = encodeForHTML(input); // Simulated
        resp.getWriter().write("<div>" + safe + "</div>");
    }

    /*
     * #F02 - OWASP ESAPI encodeForJavaScript
     * WHY SAFE: ESAPI.encoder().encodeForJavaScript() escapes for JS string context.
     *           Prevents XSS in JavaScript string literals.
     * WHY CXQL SHOULD RECOGNIZE: ESAPI JavaScript encoder is in sanitizer list.
     * TEST PURPOSE: Verify ESAPI JS encoder recognition.
     */
    public void handleEsapiJs(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("data");
        // String safe = ESAPI.encoder().encodeForJavaScript(input);
        String safe = encodeForJavaScript(input); // Simulated
        resp.getWriter().write("<script>var x = '" + safe + "';</script>");
    }

    /*
     * #F03 - OWASP ESAPI encodeForURL
     * WHY SAFE: ESAPI.encoder().encodeForURL() URL-encodes for safe URL embedding.
     *           Prevents URL injection and XSS in URL context.
     * WHY CXQL SHOULD RECOGNIZE: ESAPI URL encoder is in sanitizer list.
     * TEST PURPOSE: Verify ESAPI URL encoder recognition.
     */
    public void handleEsapiUrl(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("redirect");
        // String safe = ESAPI.encoder().encodeForURL(input);
        String safe = encodeForURL(input); // Simulated
        resp.getWriter().write("<a href=\"" + safe + "\">Link</a>");
    }

    /*
     * #F04 - Apache Commons Text escapeHtml4
     * WHY SAFE: StringEscapeUtils.escapeHtml4() escapes HTML special characters.
     *           Industry standard library used by millions of applications.
     * WHY CXQL SHOULD RECOGNIZE: Apache Commons is in CxQL's sanitizer list.
     * TEST PURPOSE: Verify Apache Commons HTML escaper recognition.
     */
    public void handleCommonsHtml(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("text");
        // String safe = StringEscapeUtils.escapeHtml4(input);
        String safe = escapeHtml4(input); // Simulated
        resp.getWriter().write("<p>" + safe + "</p>");
    }

    /*
     * #F05 - Apache Commons Text escapeXml11
     * WHY SAFE: StringEscapeUtils.escapeXml11() escapes XML special characters.
     *           Safe for XML/XHTML output contexts.
     * WHY CXQL SHOULD RECOGNIZE: Apache Commons XML escaper is in sanitizer list.
     * TEST PURPOSE: Verify Apache Commons XML escaper recognition.
     */
    public void handleCommonsXml(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("xml");
        // String safe = StringEscapeUtils.escapeXml11(input);
        String safe = escapeXml11(input); // Simulated
        resp.setContentType("application/xml");
        resp.getWriter().write("<data>" + safe + "</data>");
    }

    /*
     * #F06 - Apache Commons Text escapeEcmaScript
     * WHY SAFE: StringEscapeUtils.escapeEcmaScript() escapes for JavaScript.
     *           Prevents XSS in JavaScript string contexts.
     * WHY CXQL SHOULD RECOGNIZE: Apache Commons JS escaper should be recognized.
     * TEST PURPOSE: Verify Apache Commons JavaScript escaper recognition.
     */
    public void handleCommonsJs(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("js");
        // String safe = StringEscapeUtils.escapeEcmaScript(input);
        String safe = escapeEcmaScript(input); // Simulated
        resp.getWriter().write("<script>alert('" + safe + "');</script>");
    }

    /*
     * #F07 - Spring HtmlUtils.htmlEscape
     * WHY SAFE: Spring's HtmlUtils.htmlEscape() is the standard Spring XSS sanitizer.
     *           Used throughout Spring MVC applications.
     * WHY CXQL SHOULD RECOGNIZE: Spring HtmlUtils is in CxQL's sanitizer list.
     * TEST PURPOSE: Verify Spring HtmlUtils recognition.
     */
    public void handleSpringHtml(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("content");
        // String safe = HtmlUtils.htmlEscape(input);
        String safe = springHtmlEscape(input); // Simulated
        resp.getWriter().write("<span>" + safe + "</span>");
    }

    /*
     * #F08 - Spring HtmlUtils.htmlEscapeDecimal
     * WHY SAFE: Spring's htmlEscapeDecimal uses decimal numeric entities.
     *           Alternative encoding that's equally safe.
     * WHY CXQL SHOULD RECOGNIZE: Spring HtmlUtils variants should be recognized.
     * TEST PURPOSE: Verify Spring decimal escape recognition.
     */
    public void handleSpringDecimal(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("content");
        // String safe = HtmlUtils.htmlEscapeDecimal(input);
        String safe = springHtmlEscapeDecimal(input); // Simulated
        resp.getWriter().write("<div>" + safe + "</div>");
    }

    /*
     * #F09 - Spring HtmlUtils.htmlEscapeHex
     * WHY SAFE: Spring's htmlEscapeHex uses hexadecimal numeric entities.
     *           Alternative encoding that's equally safe.
     * WHY CXQL SHOULD RECOGNIZE: Spring HtmlUtils variants should be recognized.
     * TEST PURPOSE: Verify Spring hex escape recognition.
     */
    public void handleSpringHex(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("content");
        // String safe = HtmlUtils.htmlEscapeHex(input);
        String safe = springHtmlEscapeHex(input); // Simulated
        resp.getWriter().write("<div>" + safe + "</div>");
    }

    /*
     * #F10 - Apache Struts ResponseUtils.filter
     * WHY SAFE: Struts ResponseUtils.filter() escapes HTML special characters.
     *           Legacy but still widely used in Struts applications.
     * WHY CXQL SHOULD RECOGNIZE: Struts ResponseUtils is in sanitizer list.
     * TEST PURPOSE: Verify Struts filter recognition.
     */
    public void handleStrutsFilter(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("message");
        // String safe = ResponseUtils.filter(input);
        String safe = strutsFilter(input); // Simulated
        resp.getWriter().write("<p>" + safe + "</p>");
    }

    /*
     * #F11 - Guava HtmlEscapers
     * WHY SAFE: Guava's HtmlEscapers.htmlEscaper().escape() is Google's HTML escaper.
     *           High-quality implementation from Google's core libraries.
     * WHY CXQL SHOULD RECOGNIZE: Guava is a major framework that should be recognized.
     * TEST PURPOSE: Verify Guava HTML escaper recognition.
     */
    public void handleGuavaHtml(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("text");
        // String safe = HtmlEscapers.htmlEscaper().escape(input);
        String safe = guavaHtmlEscape(input); // Simulated
        resp.getWriter().write("<span>" + safe + "</span>");
    }

    /*
     * #F12 - JSoup clean with Safelist
     * WHY SAFE: Jsoup.clean() with Safelist removes all non-whitelisted HTML.
     *           Powerful sanitizer that allows safe HTML while blocking XSS.
     * WHY CXQL SHOULD RECOGNIZE: JSoup is a major HTML sanitization library.
     * TEST PURPOSE: Verify JSoup clean recognition.
     */
    public void handleJsoupClean(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("html");
        // String safe = Jsoup.clean(input, Safelist.basic());
        String safe = jsoupClean(input); // Simulated
        resp.getWriter().write("<div>" + safe + "</div>");
    }

    /*
     * #F13 - Java URLEncoder.encode
     * WHY SAFE: URLEncoder.encode() converts special chars to %XX sequences.
     *           Standard Java URL encoding prevents XSS in URL contexts.
     * WHY CXQL SHOULD RECOGNIZE: Java standard library URL encoder.
     * TEST PURPOSE: Verify URLEncoder recognition for URL context.
     */
    public void handleUrlEncoder(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("query");
        String safe = java.net.URLEncoder.encode(input, "UTF-8");
        resp.getWriter().write("<a href=\"search?" + safe + "\">Search</a>");
    }

    /*
     * #F14 - OWASP ESAPI encodeForHTMLAttribute
     * WHY SAFE: ESAPI encodeForHTMLAttribute() is specifically for attribute context.
     *           Handles attribute-specific XSS vectors.
     * WHY CXQL SHOULD RECOGNIZE: ESAPI attribute encoder should be recognized.
     * TEST PURPOSE: Verify ESAPI attribute encoder recognition.
     */
    public void handleEsapiAttr(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("title");
        // String safe = ESAPI.encoder().encodeForHTMLAttribute(input);
        String safe = encodeForHTMLAttribute(input); // Simulated
        resp.getWriter().write("<div title=\"" + safe + "\">Content</div>");
    }

    /*
     * #F15 - OWASP ESAPI encodeForCSS
     * WHY SAFE: ESAPI encodeForCSS() escapes for CSS context.
     *           Prevents CSS injection attacks.
     * WHY CXQL SHOULD RECOGNIZE: ESAPI CSS encoder should be recognized.
     * TEST PURPOSE: Verify ESAPI CSS encoder recognition.
     */
    public void handleEsapiCss(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String input = req.getParameter("color");
        // String safe = ESAPI.encoder().encodeForCSS(input);
        String safe = encodeForCSS(input); // Simulated
        resp.getWriter().write("<div style=\"color: " + safe + "\">Text</div>");
    }

    // ========== Simulated Framework Methods ==========
    // These simulate the actual framework methods for testing purposes
    // In production, use the real framework imports

    private String encodeForHTML(String input) {
        if (input == null) return "";
        return input.replace("&", "&amp;").replace("<", "&lt;")
                    .replace(">", "&gt;").replace("\"", "&quot;").replace("'", "&#x27;");
    }

    private String encodeForJavaScript(String input) {
        if (input == null) return "";
        return input.replace("\\", "\\\\").replace("'", "\\'")
                    .replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r");
    }

    private String encodeForURL(String input) {
        try {
            return java.net.URLEncoder.encode(input != null ? input : "", "UTF-8");
        } catch (Exception e) { return ""; }
    }

    private String escapeHtml4(String input) {
        return encodeForHTML(input);
    }

    private String escapeXml11(String input) {
        return encodeForHTML(input);
    }

    private String escapeEcmaScript(String input) {
        return encodeForJavaScript(input);
    }

    private String springHtmlEscape(String input) {
        return encodeForHTML(input);
    }

    private String springHtmlEscapeDecimal(String input) {
        return encodeForHTML(input);
    }

    private String springHtmlEscapeHex(String input) {
        return encodeForHTML(input);
    }

    private String strutsFilter(String input) {
        return encodeForHTML(input);
    }

    private String guavaHtmlEscape(String input) {
        return encodeForHTML(input);
    }

    private String jsoupClean(String input) {
        // Simulates Jsoup.clean with basic safelist
        if (input == null) return "";
        return input.replaceAll("<script[^>]*>.*?</script>", "")
                    .replaceAll("<[^>]+>", "");
    }

    private String encodeForHTMLAttribute(String input) {
        return encodeForHTML(input);
    }

    private String encodeForCSS(String input) {
        if (input == null) return "";
        return input.replaceAll("[^a-zA-Z0-9#]", "");
    }
}

